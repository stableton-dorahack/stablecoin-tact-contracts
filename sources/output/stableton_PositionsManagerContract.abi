{"name":"PositionsManagerContract","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"ChangeOwner","header":256331011,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"DepositCollateralMessage","header":299872568,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawCollateralMessage","header":1747572062,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"debtRate","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"WithdrawStablecoinMessage","header":4082453719,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"debtRate","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"RepayStablecoinMessage","header":1854010438,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"debtRate","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DoWithdrawCollateralMessage","header":3285005247,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"IncreaseTotalStableMessage","header":2517198105,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DecreaseTotalStableMessage","header":3984666698,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"MintFeesMessage","header":2638369821,"fields":[{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"JettonData","header":null,"fields":[{"name":"totalSupply","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mintable","type":{"kind":"simple","type":"bool","optional":false}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"content","type":{"kind":"simple","type":"cell","optional":true}},{"name":"walletCode","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"JettonWalletData","header":null,"fields":[{"name":"balance","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"master","type":{"kind":"simple","type":"address","optional":false}},{"name":"walletCode","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"SetUserPositionDependecyMessage","header":4030201083,"fields":[{"name":"stablecoinMasterAddress","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionsManagerAddress","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SetPositionIdMessage","header":1245051344,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"positionId","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"NewPositionIdMessage","header":225756565,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SetPositionAddressMessage","header":1571217046,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"position","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"MintMessage","header":2419415474,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayBurnMessage","header":2959459244,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"RepayBurnNotification","header":2355055729,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"StablecoinBurnedMessage","header":3929120156,"fields":[{"name":"user","type":{"kind":"simple","type":"address","optional":false}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"fees","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"TokenTransfer","header":260734629,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"destination","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseDestination","type":{"kind":"simple","type":"address","optional":true}},{"name":"customPayload","type":{"kind":"simple","type":"cell","optional":true}},{"name":"forwardTonAmount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenTransferInternal","header":395134233,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"from","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}},{"name":"forwardTonAmount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenNotification","header":1935855772,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"from","type":{"kind":"simple","type":"address","optional":false}},{"name":"forwardPayload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"TokenBurn","header":1499400124,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}}]},{"name":"TokenBurnNotification","header":2078119902,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"owner","type":{"kind":"simple","type":"address","optional":false}},{"name":"responseAddress","type":{"kind":"simple","type":"address","optional":true}}]},{"name":"TokenExcesses","header":3576854235,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"TokenUpdateContent","header":201882270,"fields":[{"name":"content","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"PositionState","header":null,"fields":[{"name":"collateral","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"debt","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]}],"receivers":[{"receiver":"internal","message":{"kind":"typed","type":"DepositCollateralMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"NewPositionIdMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawStablecoinMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"IncreaseTotalStableMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"RepayStablecoinMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"DecreaseTotalStableMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawCollateralMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"DoWithdrawCollateralMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}}],"getters":[{"name":"lastPositionId","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"balance","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"5576":{"message":"not from gateKeeper"},"15032":{"message":"not from stablecoin master"},"16059":{"message":"Invalid value"},"22230":{"message":"Already set"},"31797":{"message":"debt less than repay amount"},"53160":{"message":"not from positions manager"},"61910":{"message":"not from positionsManager"},"63577":{"message":"withdrawal amount more than position has"}}}