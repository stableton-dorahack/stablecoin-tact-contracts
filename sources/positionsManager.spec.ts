import { Address, OpenedContract, Sender, toNano } from "ton";
import { ContractSystem, Logger, Tracker, Treasure } from "ton-emulator";
import { PositionsManagerContract } from "./output/stableton_PositionsManagerContract";
import { UserPositionContract } from "./output/stableton_UserPositionContract";

describe("PositionsManagerContract", () => {
    let system: ContractSystem;
    let owner: Treasure;
    let user: Treasure;
    let gateKeeper: Treasure;
    let stablecoinMaster: Treasure;
    let positionsManagerContract: OpenedContract<PositionsManagerContract>;
    let track: Tracker;
    let logger: Logger;

    beforeAll(async () => {
        system = await ContractSystem.create();
        owner = system.treasure("owner");
        user = system.treasure("user");
        gateKeeper = system.treasure("gateKeeper");
        stablecoinMaster = system.treasure("stablecoinMaster");
        positionsManagerContract = system.open(
            await PositionsManagerContract.fromInit(gateKeeper.address, stablecoinMaster.address)
        );
        track = system.track(positionsManagerContract.address);
        logger = system.log(positionsManagerContract.address);
    });

    it("should deploy correctly", async () => {
        await positionsManagerContract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });

        await system.run();
        expect(track.collect()).toMatchSnapshot();

        expect(await positionsManagerContract.getLastPositionId()).toEqual(0n);
    });

    it("revert depositCollateral if sent not from gateKeeper", async () => {
        await positionsManagerContract.send(
            owner,
            { value: toNano(101) },
            { $$type: "DepositCollateralMessage", user: user.address, amount: toNano(100) }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("revert if collateral less than 101 ton", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(100) },
            { $$type: "DepositCollateralMessage", user: user.address, amount: toNano(100) }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("deploys new user position contract, keep collateral and forward addCollateral ", async () => {
        const balanceBefore = await positionsManagerContract.getBalance();

        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(101) },
            { $$type: "DepositCollateralMessage", user: user.address, amount: toNano(100) }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();

        const balanceAfter = await positionsManagerContract.getBalance();

        expect(balanceAfter - balanceBefore).toBeGreaterThan(BigInt(toNano(100)));
    });

    it("newPositionId request from userPosition, increase lastPositionId and send it back to sender ", async () => {
        const userPosition = system.treasure("userPosition");

        await positionsManagerContract.send(
            userPosition,
            { value: toNano(1) },
            { $$type: "NewPositionIdMessage", user: user.address }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();

        expect(await positionsManagerContract.getLastPositionId()).toBe(1n);
    });

    it("on withdrawStable from gateKeeper forwards message to the userPosition", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "WithdrawStablecoinMessage", user: user.address, amount: toNano(100), debtRate: 1n }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("on increaseTotalStable from userPosition forwards message to the gateKeeper", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "IncreaseTotalStableMessage", user: user.address, amount: toNano(100) }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
        // expect(logger.collect()).toMatchSnapshot();
    });

    it("on repayStable from gateKeeper forwards message to the userPosition", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "RepayStablecoinMessage", user: user.address, amount: toNano(100), debtRate: 1n }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("on decreaseTotalStable from userPosition forwards message to the gateKeeper", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "DecreaseTotalStableMessage", user: user.address, amount: toNano(100), fees: 1n }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
        // expect(logger.collect()).toMatchSnapshot();
    });

    it("on withdrawCollateral from gateKeeper forwards message to the userPosition", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "WithdrawCollateralMessage", user: user.address, amount: toNano(100), debtRate: 1n }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it("on doWithdrawCollateral from userPosition forwards message to the gateKeeper", async () => {
        await positionsManagerContract.send(
            gateKeeper,
            { value: toNano(1) },
            { $$type: "DoWithdrawCollateralMessage", user: user.address, amount: toNano(100) }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot();
        // expect(logger.collect()).toMatchSnapshot();
    });
});
