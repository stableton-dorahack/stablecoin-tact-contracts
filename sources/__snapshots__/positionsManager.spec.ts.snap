// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionsManagerContract deploys new user position contract and forward addCollateral  1`] = `
[
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{DDED1FE8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 101000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 22548n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{DDED1FE8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
            "type": "internal",
            "value": 100942928000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{0D74C595800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B3_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 98601000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20542n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{4A35F9D0800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000000000000000000000000000000000000000000000000000000000000001}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
            "type": "internal",
            "value": 98344000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract newPositionId request from userPosition, increase lastPositionId and send it back to sender  1`] = `
[
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{0D74C595800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B3_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBqJ1A2wQ5MkOBlhdTr7ML850tmbmDecFDAEvNUwHvm0sGh",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20542n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{4A35F9D0800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000000000000000000000000000000000000000000000000000000000000002}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQBqJ1A2wQ5MkOBlhdTr7ML850tmbmDecFDAEvNUwHvm0sGh",
            "type": "internal",
            "value": 98344000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on decreaseTotalStable from userPosition forwards message to the gateKeeper 1`] = `
[
  {
    "$seq": 10,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{ED81304A800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD0000000000000000003_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 21034n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{ED81304A800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD0000000000000000003_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
            "type": "internal",
            "value": 98473000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on doWithdrawCollateral from userPosition forwards message to the gateKeeper 1`] = `
[
  {
    "$seq": 11,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C3CD33BF800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 21182n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C3CD33BF800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD001_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
            "type": "internal",
            "value": 98537000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on increaseTotalStable from userPosition forwards message to the gateKeeper 1`] = `
[
  {
    "$seq": 7,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{96096519800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD001_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20689n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{96096519800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD001_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
            "type": "internal",
            "value": 977848000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on repayStable from gateKeeper forwards message to the userPosition 1`] = `
[
  {
    "$seq": 8,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{BEF31E07800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 21848n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{BEF31E07800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
            "type": "internal",
            "value": 98309000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 9,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 91178000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 6330n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on withdrawStable from gateKeeper forwards message to the userPosition 1`] = `
[
  {
    "$seq": 5,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQBhDJZIy0_RS_IdyGG5GxdsC8eWb15DS9b3IL2p2-b0cUi8",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 21653n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000002E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
            "to": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
            "type": "internal",
            "value": 976656000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 6,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQD-q_URnMcLNDmDEG6mn-oUsoeIHbzZCmHSBKok9925RPb1",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 958770000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 6330n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract revert depositCollateral if sent not from gateKeeper 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{DDED1FE8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 101000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 5576,
        "errorMessage": "not from gateKeeper",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "type": "internal",
          "value": 100993705000n,
        },
      },
    ],
  },
]
`;

exports[`PositionsManagerContract should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{23071186801075E834DF4F15E107E94B1EDF44C910774A333E3C6B7B2F37E39B16FED191AD9001843259232D3F452FC8772186E46C5DB02F1E59BD790D2F5BDC82F6A76F9BD1C6_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQDABUOp56Z0-s0ottjM90287XtNhXKPNbz9LlpHIqdmaNAq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 9256n,
      },
    ],
  },
]
`;
